File: 백준/Silver/24460. 특별상이라도 받고 싶어/특별상이라도 받고 싶어.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine());

        int[][] arr = new int[n][n];

        for (int i = 0; i < n; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            for (int j = 0; j < n; j++) {
                arr[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        System.out.println(solve(n, 0, 0, arr));
    }

    private static int solve(int n, int x, int y, int[][] arr) {

        if (n == 1) {
            return arr[x][y];
        }

        int[] temp = new int[4];

        int half = n / 2;

        temp[0] = solve(half, x, y, arr);
        temp[1] = solve(half, x, y + half, arr);
        temp[2] = solve(half, x + half, y, arr);
        temp[3] = solve(half, x + half, y + half, arr);

        Arrays.sort(temp);

        return temp[1];
    }
}

File: 백준/Silver/24460. 특별상이라도 받고 싶어/특별상이라도 받고 싶어.py
def solve(n, x, y):
    if n == 1:
        return arr[x][y]

    half = n // 2

    temp = [
        solve(half, x, y),
        solve(half, x, y + half),
        solve(half, x + half, y),
        solve(half, x + half, y + half)
    ]

    temp.sort()

    return temp[1]


n = int(input())
arr = [list(map(int, input().split())) for _ in range(n)]

print(solve(n, 0, 0))
